File: src/agents/prompt/builder.ts
import type { RetrievedChunk } from '../ports/knowledge.port.js';
import type { Fact } from '../ports/memory.port.js';
import type { StoredMessage } from '../domain/message.js';
import { SYSTEM_PROMPT } from './system.js';
import { trimToBudget } from './budget.js';

/**
 * Thu tu: thong tin ON DINH truoc, thong tin TUOI sau.
 * Vua hop co che chu y cua model, vua tan dung prompt caching cho phan dau it doi.
 * Vi tri: system -> L4 -> L3 -> L2 -> L1 -> cau hoi.
 */

/**
 * Chan noi dung tu thoat khoi hop bang cach viet the dong cua chinh no.
 *
 * Khong lam buoc nay thi mot tai lieu chua `</tai_lieu>` se tu thoat ra khoi hop —
 * va do chinh xac la cach nguoi ta pha. Ap cho CA ket qua cong cu (tools/guard.ts),
 * noi rui ro cao hon nhieu vi van ban do nguoi la soan.
 */
export function sanitize(content: string): string {
  return content.replace(/<\/?(tai_lieu|ket_qua_cong_cu|ghi_nho)[^>]*>/gi, '');
}

export function renderKnowledge(chunks: readonly RetrievedChunk[]): string {
  if (chunks.length === 0) return '';
  const body = chunks
    .map(
      (c) =>
        `<tai_lieu id="${c.chunkId}" nguon="${c.docTitle}" muc="${c.section ?? ''}">\n` +
        `${sanitize(c.content)}\n</tai_lieu>`,
    )
    .join('\n');
  return trimToBudget(body, 'knowledge').text;
}

export function renderFacts(facts: readonly Fact[]): string {
  if (facts.length === 0) return '';
  const body = `<ghi_nho>\n${facts.map((f) => `- ${f.content}`).join('\n')}\n</ghi_nho>`;
  return trimToBudget(body, 'facts').text;
}

export function renderRecent(messages: readonly StoredMessage[], isGroup: boolean): string {
  const body = messages
    .map((m) => (isGroup ? `[${m.senderName}]: ${m.text}` : m.text))
    .join('\n');
  return trimToBudget(body, 'recent').text;
}

export { SYSTEM_PROMPT };

File: src/agents/prompt/builder.py
"""Render tung khoi ngu canh thanh van ban dua vao prompt."""

import re

from ..domain.knowledge import RetrievedChunk
from ..domain.message import StoredMessage
from ..ports.memory import Fact
from .budget import trim_to_budget

_TAG_LOOKALIKE = re.compile(r"</?(tai_lieu|ket_qua_cong_cu|ghi_nho)[^>]*>", re.IGNORECASE)


def sanitize(content: str) -> str:
    """Chan noi dung tu thoat khoi hop bang cach viet the dong cua chinh no.

    Khong lam buoc nay thi mot tai lieu chua `</tai_lieu>` se tu thoat ra khoi hop —
    va do chinh xac la cach nguoi ta pha. Ap cho CA ket qua cong cu (tools/guard.py),
    noi rui ro cao hon nhieu vi van ban do nguoi la soan.
    """
    return _TAG_LOOKALIKE.sub("", content)


def render_knowledge(chunks: tuple[RetrievedChunk, ...]) -> str:
    if not chunks:
        return ""
    body = "\n".join(
        f'<tai_lieu id="{c.chunk_id}" nguon="{c.doc_title}" muc="{c.section or ""}">\n'
        f"{sanitize(c.content)}\n</tai_lieu>"
        for c in chunks
    )
    return trim_to_budget(body, "knowledge").text


def render_facts(facts: tuple[Fact, ...]) -> str:
    if not facts:
        return ""
    body = "<ghi_nho>\n" + "\n".join(f"- {f.content}" for f in facts) + "\n</ghi_nho>"
    return trim_to_budget(body, "facts").text


def render_recent(messages: tuple[StoredMessage, ...], is_group: bool) -> str:
    body = "\n".join(f"[{m.sender_name}]: {m.text}" if is_group else m.text for m in messages)
    return trim_to_budget(body, "recent").text
